flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSourceSinkFactoryBase.java

```diff
@@ -314,9 +322,24 @@ private StartupOptions getStartupOptions(
 		return options;
 	}
 
-	private FlinkKafkaPartitioner<Row> getFlinkKafkaPartitioner() {
-		// we don't support custom partitioner so far
-		return new FlinkFixedPartitioner<>();
+	@SuppressWarnings("unchecked")
+	private Optional<FlinkKafkaPartitioner<Row>> getFlinkKafkaPartitioner(DescriptorProperties descriptorProperties) {
+		return descriptorProperties
+			.getOptionalString(CONNECTOR_SINK_PARTITIONER)
+			.flatMap((String partitionerString) -> {
+				switch (partitionerString) {
```

Comment:
It seems like there is a missing case here: key-hash partitioning.

That would be the default partitioning scheme used in the Kafka producer, if 1) a key exists for records (returned by the `KeyedDeserializationSchema`), and 2) there is no partitioner provided.

Round-robin partitioning only takes place if there are no record keys.

Reply:
Currently, we don't support the definition of keys and thus we don't support `KeyedDeserializationSchema`. I will open a follow-up issue for this. I guess it would require a new type of format that takes two nested formats for `key` and `value`.